File: python/nwalign/fasta.py

import gzip


class FastaRecord(object):

    def __init__(self, id_, sequence, description=None):
        self.id = id_
        self.sequence = sequence
        self.description = description

    @property
    def header(self):
        if not self.description:
            return "".join([">", self.id])
        else:
            return "".join([">", self.id, " ", self.description])

    def format(self, width=60):
        lines = [self.header]
        k = 0
        while k < len(self.sequence):
            lines.append(self.sequence[k:(k + width)])
            k += width
        return "\n".join(lines)

    def __str__(self):
        return self.format()

    def __getitem__(self, idx):
        return self.sequence[idx]

    def __len__(self):
        return len(self.sequence)


def read_fasta(file_name):
    _open_file = lambda f: open(f, "rU")
    if file_name.endswith(".gz"):
        _open_file = lambda f: gzip.open(f, "r")
    records = []

    def new_record(h, s):
        fields = h.lstrip(">").strip().split()
        if len(fields) == 0:
            raise IOError("Cannot parse as FASTA header:\n{h}".format(h=l))
        elif len(s) == 0:
            raise IOError("Empty sequence for FASTA record:\n{h}".format(h=l))
        return FastaRecord(fields[0], "".join(s), " ".join(fields[1:]))
    current_header = None
    current_sequence = None
    with _open_file(file_name) as f:
        for line in f:
            if line.strip() == "":
                raise IOError("Empty lines not allowed")
            elif line.startswith(">"):
                if current_header is not None:
                    records.append(
                        new_record(current_header, current_sequence))
                current_header = line
                current_sequence = []
            else:
                current_sequence.append(line.strip())
    if current_header is None or current_sequence is None:
        raise IOError("No valid FASTA records in {f}".format(f=file_name))
    records.append(new_record(current_header, current_sequence))
    return records

File: rust/src/fasta.rs
// Author: Nat Echols
// License: public domain

use std::fs::File;
use std::io::{BufRead,BufReader};
use std::path::Path;

/// Representation of a FASTA file record, with header split into ID and
/// description
pub struct FastaRecord {
  pub id: String,
  pub sequence: String,
  pub description: Option<String>,
}

impl FastaRecord {

  /// Return the full record header including description, if any
  pub fn header(&self) -> String {
    match self.description {
      Some(ref desc) => format!(">{} {}", self.id, desc),
      None => format!(">{}", self.id)
    }
  }

  // TODO
  //pub fn format(&self, width: u32) -> String {
  //}
}

fn new_record(header: String, sequence_lines: Vec<String>) -> FastaRecord {
  if sequence_lines.len() == 0 {
    panic!("Empty sequence for {}", header);
  } else if header.len() == 0 {
    panic!("Empty header line not allowed");
  }
  let mut header_fields = header.split_whitespace();
  let id = header_fields.next().unwrap().to_string();
  let mut desc_fields = Vec::with_capacity(1);
  loop {
    let field = header_fields.next();
    match field {
      Some(s) => desc_fields.push(s),
      None => break
    };
  }
  let desc = if desc_fields.len() > 0 {
    Some(desc_fields.join(" "))
  } else {
    None
  };
  FastaRecord{id: id, sequence: sequence_lines.join(""), description: desc}
}

/// Read a FASTA file line by line and return a Vec of FastaRecord objects
pub fn read_fasta(file_name: &str) -> Vec<FastaRecord> {
  let file = match File::open(Path::new(file_name)) {
    Err(_) => panic!("Couldn't open {}", file_name),
    Ok(f) => f
  };
  let mut records = Vec::with_capacity(1);
  let mut current_header: Option<String> = None;
  let mut current_sequence: Vec<String> = Vec::with_capacity(1);
  for line in BufReader::new(file).lines() {
    let l = line.unwrap();
    if l.starts_with(">") {
      if current_header.is_some() {
        records.push(new_record(current_header.unwrap(), current_sequence));
      }
      current_header = Some(l.trim_left_matches('>').trim().to_string());
      current_sequence = Vec::with_capacity(1);
    } else if current_header.is_some() {
      current_sequence.push(l.trim().to_string());
    } else {
      panic!("Illegal FASTA line without header:\n{}", l);
    }
  }
  if current_header.is_none() {
    panic!("No valid FASTA records in file");
  }
  records.push(new_record(current_header.unwrap(), current_sequence));
  records
}


#[cfg(test)]
mod tests {
  //use tempdir::TempDir;
  use super::read_fasta;
  use std::io::Write;
  use std::fs;

  #[test]
  fn test_fasta() {
    // FIXME figure out how to use temp files properly
    //let tmp_dir = TempDir::new("test-fasta");
    //let tmp_file = tmp_dir.path().join("test1.fasta");
    let mut f = fs::File::create("test1.fasta");
    writeln!(f.unwrap(), ">seq1\nHEAGAWGHEE\n>seq2 my sequence\nPAWHEAE");
    let records = read_fasta("test1.fasta");
    assert_eq!(records.len(), 2);
    assert_eq!(records[0].id, "seq1".to_string());
    assert_eq!(records[0].sequence, "HEAGAWGHEE".to_string());
    assert!(records[0].description.is_none());
    assert_eq!(records[0].header(), ">seq1".to_string());
    assert_eq!(records[1].id, "seq2".to_string());
    assert_eq!(records[1].sequence, "PAWHEAE".to_string());
    assert!(records[1].description.is_some());
    assert_eq!(records[1].header(), ">seq2 my sequence".to_string());
    fs::remove_file("test1.fasta");
  }

  // TODO test failure modes
}
